src/main/java/org/hmhb/program/DefaultProgramService.java

```diff
@@ -69,16 +74,169 @@ public Program getById(
     ) {
         LOGGER.debug("getByIds called: ids={}", ids);
         requireNonNull(ids, "ids cannot be null");
-        return dao.findByIdIn(ids);
+        return dao.findAll(ids);
     }
 
     @Timed
     @Override
     public List<Program> getAll() {
         LOGGER.debug("getAll called");
-        List<Program> target = new ArrayList<>();
-        dao.findAll().forEach(target::add);
-        return target;
+        return dao.findAllByOrderByNameAsc();
+    }
+
+    @Timed
+    @Override
+    public List<Program> search(
+            @Nonnull MapQuery query
+    ) {
+        LOGGER.debug("search called: query={}", query);
+        requireNonNull(query, "query cannot be null");
+
+        MapQuerySearch search = query.getSearch();
+
+        if (search == null) {
+            return getAll();
+        }
+
+        if (search.getProgram() != null) {
+
+            return Collections.singletonList(
+                    getById(search.getProgram().getId())
+            );
+
+        }
+
+        County county = search.getCounty();
+        Organization organization = search.getOrganization();
+        ProgramArea programArea = search.getProgramArea();
+
+        Long countyId = null;
+
+        if (county != null) {
+            countyId = county.getId();
+        }
+
+        Long organizationId = null;
+
+        if (organization != null) {
+            organizationId = organization.getId();
+        }
+
+        Long programAreaId = null;
+
+        if (programArea != null) {
+            programAreaId = programArea.getId();
+        }
+
+        /*
+         * Both of these potential implementations (nested ifs and flattened ifs) are really ugly,
+         * but using JPA Specification looks much more complicated.
+         *
+         * For now, with only 3 search params, I'm just going to have all permutations of
+         * searches in the DAO.  If anymore params are added and the permutations get worse,
+         * I'll have to use JPA Specification or figure something else out.
+         */
+//        return doSearchFlattenedIfs(organizationId, countyId, programAreaId);
```

Comment:
I don't really like either of these implementations.  @mckoon please review them and give a preference on which to keep.

The alternative (JPA Specification) seemed much worse for our needs (at least at the moment):

https://www.petrikainulainen.net/programming/spring-framework/spring-data-jpa-tutorial-part-four-jpa-criteria-queries/
http://docs.spring.io/spring-data/jpa/docs/current/api/org/springframework/data/jpa/domain/Specifications.html

There was something mentioned called Querydsl that I didn't look into much, but I didn't want to drag in another tech/library:

https://spring.io/blog/2011/04/26/advanced-spring-data-jpa-specifications-and-querydsl/